File: tools/manual/run_ht.cpp
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT license.

// Largely derived from https://github.com/microsoft/ADBench/blob/38cb7931303a830c3700ca36ba9520868327ac87/src/cpp/modules/manual/ManualHand.cpp

#include "gradbench/main.hpp"
#include "gradbench/evals/ht.hpp"

#include "ht_d.hpp"

class Jacobian : public Function<ht::Input, ht::JacOutput> {
  bool _complicated = false;
  std::vector<double> _objective;
public:
  Jacobian(ht::Input& input) :
    Function(input),
    _complicated(input.us.size() != 0),
    _objective(3 * input.data.correspondences.size())
  {}

  void compute(ht::JacOutput& output) {
    int err_size = 3 * _input.data.correspondences.size();
    int ncols = (_complicated ? 2 : 0) + _input.theta.size();
    output.jacobian_ncols = ncols;
    output.jacobian_nrows = err_size;
    output.jacobian.resize(err_size * ncols);

    if (_complicated) {
      ht_objective_d(_input.theta.data(), _input.us.data(), _input.data, _objective.data(), output.jacobian.data());
    } else {
      ht_objective_d(_input.theta.data(), _input.data, _objective.data(), output.jacobian.data());
    }
  }
};

int main(int argc, char* argv[]) {
  return generic_main(argc, argv, {
      {"objective", function_main<ht::Objective>},
      {"jacobian", function_main<Jacobian>}
    });
}

File: tools/zygote-jl/run_ht.jl
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT license.
#
# Incorporates code from: https://github.com/microsoft/ADBench/blob/38cb7931303a830c3700ca36ba9520868327ac87/src/julia/modules/Zygote/ZygoteHT.jl

module HT

import Zygote
import GradBench

# FIXME: it is very expensive to redo all the input parsing here for
# every run. We absolutely must hoist it out into a "prepare" stage.
function objective(j)
    input = GradBench.HT.input_from_json(j)
    complicated = size(input.us, 1) != 0
    if complicated
        GradBench.HT.objective_complicated(input.model,
                                           input.correspondences,
                                           input.points,
                                           input.theta,
                                           input.us)
    else
        GradBench.HT.objective_simple(input.model,
                                      input.correspondences,
                                      input.points,
                                      input.theta)
    end
end

function jacobian(j)
    input = GradBench.HT.input_from_json(j)
    complicated = size(input.us, 1) != 0
    if complicated
        wrapper = theta -> GradBench.HT.objective_complicated(input.model, input.correspondences, input.points, theta, input.us)
        y, jacobian_theta = Zygote.forward_jacobian(wrapper, input.theta)
        ylen = size(y, 1)
        jacobian_us = hcat([
            begin
                _, ju = Zygote.forward_jacobian(u -> GradBench.HT.objective_complicated(input.model, input.correspondences, input.points, input.theta, vcat(input.us[1:j-1], [u], input.us[j+1:end])), input.us[j])
                ju[:, 3j-2:3j]
            end
            for j ∈ 1:size(input.us, 1)
                ]...)
        vcat(jacobian_us, jacobian_theta)
    else
        Zygote.forward_jacobian(theta -> GradBench.HT.objective_simple(input.model, input.correspondences, input.points, theta), input.theta)[2]
    end
end

GradBench.register!("ht", Dict(
    "objective" => objective,
    "jacobian" => jacobian
))

end
